### PR TITLE
Multi gpu support

### PR DESCRIPTION
**Pull request by DeNeutoy**
This PR adds data parallelism which allows batch updates to be split across any number of GPUs with synchronous gradient updates performed on the CPU. 

Additionally, I've added a tensorflow device scope to the `_embed_input` function, which should locate all embedding variables on the CPU. 

There are a few things to decide here:
1) Default behaviour when the number of available GPUS is not what it is in the model parameters. At the moment (and by default in Keras), we enable `soft_device_placement=True` in the session configuration, which means that tensorflow will try to put things on specified GPUs, but won't complain if they aren't available and will just allocate them some other way if the devices aren't there.

2) Should we keep the device_placement logging/make it a parameter? It will print all of the operations in the graph and what device they are on, which is a lot of stuff, but also extremely helpful for debugging etc.

### CODE REVIEW

**Comment by matt-peters** [comment]
Does this method work and give performance speedups?  E.g. if you run on multiple GPUs and manually inspect nvidia-smi, are all GPUs working equally hard?  Or how much speedup do you get vs just using a larger batch on a single GPU?

It looks like you are creating only one copy of each variable/weight on a single device.  Then multiple input and output tensors separated across devices.  I'm curious what tensorflow would do in this situation:

* copy weights to each GPU to do computation
* copy the input tensors to the device with the weight matrices and do computation there
* something else?

**Comment by DeNeutoy** [comment]
Yeah - I think this is copying weights back and forth to the different GPUs, with the speedup being that this can be done in parallel - however, I need to look into the difference between this and using one of tensorflow's parameter servers, as that seems to be how they want people to do this distributed stuff...

File: deep_qa/training/text_trainer.py [hunk 1]
```diff
@@ -305,7 +306,13 @@ def _embed_input(self, input_layer: Layer, embedding_name: str="embedding"):
         uses both words and characters, we need to run the character encoder and concatenate the
         result with a word embedding).
         """
-        return self.tokenizer.embed_input(input_layer, self.__get_embedded_input, self, embedding_name)
+        with tensorflow.device("/cpu:0"):
```
Comment (matt-gardner) [inline deep_qa/training/text_trainer.py]: I think this is the wrong place for this.  If you're using a `word and characters` tokenizer, `tokenizer.embed_input` gets a word embedding and also does a character-level CNN, then concatenates them.  You only want the embedding lookups to happen on the CPU, not the CNN.  I also am thinking that MattP's pre-trained language model stuff will fit into DeepQA in the call to `self._embed_input`, so this could also get a lot more complex in some scenarios.  I think the right place to put the device scope is [here](https://github.com/allenai/deep_qa/blob/00d36306759cb1c232489f68844371fb727ce2c8/deep_qa/training/text_trainer.py#L602).

File: deep_qa/training/multi_gpu.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from keras.layers import merge
+from keras.layers.core import Lambda
+from .models import DeepQaModel
+import tensorflow as tf
+
+
+def make_parallel(model: DeepQaModel, gpu_count: int) -> DeepQaModel:
+
+    """
+    :param model: An instance of a DeepQaModel.
```
Comment (matt-gardner) [inline deep_qa/training/multi_gpu.py]: We're using a different docstring format.  This won't work with our doc tools (and even if I'm wrong and it does, we should still just standardize on a single docstring format, for consistency and readability across the library).

File: deep_qa/training/text_trainer.py (hunk 1 above)
Comment (DeNeutoy) [inline deep_qa/training/text_trainer.py]: Ah yep, I forgot about the cnn encoder stuff. Thanks.
